pkg/bloomgateway/bloomgateway.go

```diff
@@ -99,18 +194,116 @@ func New(cfg Config, schemaCfg config.SchemaConfig, storageCfg storage.Config, s
 	}
 
 	g.bloomStore = bloomStore
-	g.Service = services.NewIdleService(g.starting, g.stopping)
+
+	svcs := []services.Service{g.queue, g.activeUsers}
+	g.serviceMngr, err = services.NewManager(svcs...)
+	if err != nil {
+		return nil, err
+	}
+	g.serviceWatcher = services.NewFailureWatcher()
+	g.serviceWatcher.WatchManager(g.serviceMngr)
+
+	g.Service = services.NewBasicService(g.starting, g.running, g.stopping).WithName("bloom-gateway")
 
 	return g, nil
 }
 
 func (g *Gateway) starting(ctx context.Context) error {
+	var err error
+	defer func() {
+		if err == nil || g.serviceMngr == nil {
+			return
+		}
+		if err := services.StopManagerAndAwaitStopped(context.Background(), g.serviceMngr); err != nil {
+			level.Error(g.logger).Log("msg", "failed to gracefully stop bloom gateway dependencies", "err", err)
+		}
+	}()
+
+	if err := services.StartManagerAndAwaitHealthy(ctx, g.serviceMngr); err != nil {
+		return errors.Wrap(err, "unable to start bloom gateway subservices")
+	}
+
+	for i := 0; i < numWorkers; i++ {
+		go g.startWorker(ctx, fmt.Sprintf("worker-%d", i))
+	}
+
 	return nil
 }
 
+func (g *Gateway) running(ctx context.Context) error {
+	// We observe inflight tasks frequently and at regular intervals, to have a good
+	// approximation of max inflight tasks over percentiles of time. We also do it with
+	// a ticker so that we keep tracking it even if we have no new requests but stuck inflight
+	// tasks (eg. worker are all exhausted).
+	inflightTasksTicker := time.NewTicker(250 * time.Millisecond)
+	defer inflightTasksTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-g.serviceWatcher.Chan():
+			return errors.Wrap(err, "bloom gateway subservice failed")
+		case <-inflightTasksTicker.C:
+			g.pendingTasks.RLock()
+			inflight := len(g.pendingTasks.Map)
+			g.pendingTasks.RUnlock()
+			g.metrics.inflightRequests.Observe(float64(inflight))
```

Comment:
Q: I'm bit confused with how we calculate `inflightRequests` metric here. From the description, I see it's combination of requests both waiting and getting processed. Why not just increment when creating and successful enqueue in `FilterChunksRefs` method and decrement after dequeue and processed on the worker loop?

why need in-flights ticker here?

> We also do it with
// a ticker so that we keep tracking it even if we have no new requests but stuck inflight
// tasks (eg. worker are all exhausted).

I understand we are calculating `inflight` requests based on `pendingTasks` shared map. And in that way, once task is dequeue and it removed from pendingTasks but if it get stuck on fetching it from BloomStore then inflights metrics can go invalid. But do we have to that way?

Instead why not increament `inflight` metrics during creation + successful enqueue (in FilterChunkRefs) and decrement it only after dequeue + success processing (after getting response from BloomStore)

Am i missing anything here?

Reply:
Afaik, the idea of using a Summary for the inflight requests is that we can better observer spikes. If we'd use a Gauge and do Inc()/Dec() for every task, we can easily miss spikes in between the 15s scrape interval.

The reasoning is also explained in the [PR](https://github.com/grafana/loki/pull/5658) of the original implementation in the query scheduler.